src/cargo/core/resolver/mod.rs

```diff
@@ -987,6 +987,10 @@ fn activate_deps_loop(
         Vec<HashMap<PackageId, ConflictReason>>,
     > = HashMap::new();
 
+    // `past_conflict_triggers` is an inverse-index of `past_conflicting_activations`.
+    // For every `PackageId` this lists the `Dependency`s that mention it in `past_conflicting_activations`.
+    let mut past_conflict_triggers: HashMap<PackageId, HashSet<Dependency>> = HashMap::new();
```

Comment:
Since this function is already pretty tricky, perhaps this  along with `past_conflicting_activations` could be moved to a separate data structure so they're managed in sync?

Reply:
Working on it, should be done tomorrow some time.